### PR TITLE
FINERACT-2060: Validation rules for loan reamortization

### PR DESCRIPTION
**Pull request by galovics**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java [hunk 1]
```diff
@@ -18,18 +18,118 @@
...
+        }
+
+        // validate reamortization is only done on an active loan
+        if (!loan.getStatus().isActive()) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.supported.only.for.active.loans",
+                    "Loan reamortization can only be done on active loans", loan.getId());
+        }
+
+        // validate if there's already a re-amortization transaction for today
+        boolean isReAmortizationTransactionForTodayPresent = !loan
```
Comment (reluxa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java]: What if the re-amortization was marked with undo?

E.g.: 

- reamoritzation on 2024.03.11
- reamortization on 2024.03.11 was undoed
- reamoritzation on 2024.03.11 <- should this be allowed?

**Review by reluxa — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java [hunk 2]
```diff
@@ -18,18 +18,118 @@
...
+        if (optionalReAmortizationTx.isEmpty()) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.reamortization.transaction.missing",
+                    "Undoing a reamortization can only be done if there was a reamortization already", loan.getId());
+        }
+
+        // validate if there's no payment between the reamortization and today
+        LocalDate reAmortizationTxDate = optionalReAmortizationTx.get().getTransactionDate();
+        boolean repaymentExistsAfterReAmortization = !loan
+                .getLoanTransactions(
+                        tx -> tx.getTypeOf().isRepaymentType() && DateUtils.isAfter(reAmortizationTxDate, getBusinessLocalDate()))
```
Comment (reluxa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java]: What if the repayment is on the same day, but it happens after the reamortization?

**Comment by reluxa** [comment]
It would be nice to see some unit tests.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java (hunk 2 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java]: Edge case, doesn't need to be handled for now.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java (hunk 1 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java]: Same day, we can't differentiate and it's an edge case.

**Review by taskain7 — APPROVED** [review]
LGTM
We can merge when you fixed the failing test cases

**Review by reluxa — APPROVED** [review]
LGTM
